### ISSUE TITLE
About the SAD in ENet-label-torch

### ISSUE THREAD

**Issue by freshfruit930** [cardwing/Codes-for-Lane-Detection#208]
After studying the paper and the model, I have several questions about the SAD and the training process:
1. As mentioned in the paper that SAD is added in the 40K episodes. Does this mean that the training process need to have two stages, first is use the model without SAD and the second use the model with SAD?
2. from the model definition, seems that there are only 2 attention maps added, not as shown as in Fig.3 in the paper, do I understand right?
3. I am still confusing on how the attention maps improves the result. Just as shown in Fig.3 in the paper, seems that the attention maps not contribute on the final result. Meanwhile, as in the code, just concatenate the attention map with the output from stage 2 and 4(if I understand right) seems also not affect the original data flow. Please help me on clarifying this.

Thanks!

**Comment by cardwing** [comment]
1. Yes.
2. Yes. As to other mimicking path, you need to add it by yourself.
3. SAD improves the representation learning of the model and hence it improves the prediction. The code in ENet-Label-Torch contains the process of block 3 mimicking block 4 and block 4 mimicking the label.

**Comment by freshfruit930** [comment]
Hi,
Thanks for your reply.
As for the third question, does block 4 means the decoder block? Seems from the model definition, the block 3 and block 4 are combined with one attention map?
Another question is the last attention map is the one as you mention that is the block 4 mimicking the label?

**Comment by rodrigoberriel** [comment]
@cardwing I have a similar question. I have no experience with Lua, but I have experience with PyTorch and other frameworks. I don't see "block 3 mimicking block 4" in the code. If I understood correctly, and maybe I didn't, this code: 

https://github.com/cardwing/Codes-for-Lane-Detection/blob/a224eb4d38cf3294ea80b2d4181f8afe028d801e/ENet-Label-Torch/train.lua#L71-L80

seems to be applying:
- Spatial CE between 1st output and `target[1]`
- Binary CE between 2nd output and `target[2]`
- MSE between 3rd output and `target[1]`
- MSE between 4th output and `target[1]` 

as defined in:

https://github.com/cardwing/Codes-for-Lane-Detection/blob/a224eb4d38cf3294ea80b2d4181f8afe028d801e/ENet-Label-Torch/models/init.lua#L158-L162

Because of my lack of experience with Lua, I don't know where to find in the code what `target` is, but based on the loss definition, `target[1]` is the segmentation map and `target[2]` is the existence ground-truth. **In that case, it seems to me that both SAD modules are trying to mimic the label.** Is that correct?

I am trying to plug SAD modules to ENet-**TuSimple**-Torch to retrain and get the model with 96.64%, but it seems I won't have it done before my deadline and I won't be able to include your method in the comparison :(

**Comment by cardwing** [comment]
That's true. I leverage deep supervision to finetune the performance of ENet-SAD and that's why the F1-measure of the provided code is better than the reported F1-measure on CULane (72.0 v.s. 70.8). As to TuSimple, since many baselines have so strong performance, several engineering tricks (e.g., warm up pre-training, lane smoothing and lane postprocessing) have been utilized to make ENet-SAD have comparable performance. You can have a test of the ENet model on TuSimple. It can achieve better result (Accuracy: 0.9486756915, FP: 0.0457901133, FN: 0.0386712197) than the reported performance (0.9302 accuracy). Reproducing the performance of ENet-SAD on TuSimple is a bit difficult since several engineering tricks are not included in the current code. If you have limited time and resources, you may just include the performance of ENet which is an essential backbone of our method.

**Comment by lucastabelini** [comment]
Could you give us more details on what kinds of "engineering tricks" were used?

**Comment by cardwing** [comment]
Like utilize image cropping to remove areas without lanes, use a smaller learning rate to pretrain the model, replace the conventional convolution with dilated convolution, cut the predicted lanes shorter if they are too long, smooth the predictions if necessary, etc.

**Comment by martijndeheus** [comment]
> @cardwing I have a similar question. I have no experience with Lua, but I have experience with PyTorch and other frameworks. I don't see "block 3 mimicking block 4" in the code. If I understood correctly, and maybe I didn't, this code:
> 
> https://github.com/cardwing/Codes-for-Lane-Detection/blob/a224eb4d38cf3294ea80b2d4181f8afe028d801e/ENet-Label-Torch/train.lua#L71-L80
> 
> seems to be applying:
> 
> * Spatial CE between 1st output and `target[1]`
> * Binary CE between 2nd output and `target[2]`
> * MSE between 3rd output and `target[1]`
> * MSE between 4th output and `target[1]`
> 
> as defined in:
> 
> https://github.com/cardwing/Codes-for-Lane-Detection/blob/a224eb4d38cf3294ea80b2d4181f8afe028d801e/ENet-Label-Torch/models/init.lua#L158-L162
> 
> Because of my lack of experience with Lua, I don't know where to find in the code what `target` is, but based on the loss definition, `target[1]` is the segmentation map and `target[2]` is the existence ground-truth. **In that case, it seems to me that both SAD modules are trying to mimic the label.** Is that correct?
> 
> I am trying to plug SAD modules to ENet-**TuSimple**-Torch to retrain and get the model with 96.64%, but it seems I won't have it done before my deadline and I won't be able to include your method in the comparison :(

Hi! I am working on implementing the SAD in Pytorch for CULane. We have similar questions to how the SAD works. This question specifically:

```Because of my lack of experience with Lua, I don't know where to find in the code what `target` is, but based on the loss definition, `target[1]` is the segmentation map and `target[2]` is the existence ground-truth. **In that case, it seems to me that both SAD modules are trying to mimic the label.** Is that correct?```

Have you by any chance figured out how this works?

@cardwing @rodrigoberriel

**Comment by cardwing** [comment]
You can just replace the target label with the output of block 3 and block 4. You can also add deep supervision to the output of block 4 to further boost the performance.
